server/command.go

```diff
@@ -85,20 +99,64 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 	return resp, nil
 }
 
+func (p *Plugin) runSendCommand(args []string, extra *model.CommandArgs) (*model.CommandResponse, bool, error) {
+	if len(args) < 2 {
+		return getCommandResponse(model.COMMAND_RESPONSE_TYPE_EPHEMERAL, "You must specify a user and a message."), false, nil
+	}
+
+	receiver, err := p.API.GetUserByUsername(args[0][1:])
+	if err != nil {
+		return getCommandResponse(model.COMMAND_RESPONSE_TYPE_EPHEMERAL, "Please, provide a valid user."), false, nil
+	}
+
+	if receiver.Id == extra.UserId {
+		return p.runAddCommand(args[1:], extra)
+	}
+
+	message := strings.Join(args[1:], " ")
+
+	senderItem := newItem(message)
+	p.storeItem(senderItem)
+	receiverItem := newItem(message)
+	p.storeItem(receiverItem)
+
+	senderList := p.getOutListForUser(extra.UserId)
+	receiverList := p.getInListForUser(receiver.Id)
+
+	appErr := senderList.add(senderItem.ID, receiverItem.ID, receiver.Id)
+	if appErr != nil {
+		return nil, false, appErr
+	}
+
+	appErr = receiverList.add(receiverItem.ID, senderItem.ID, extra.UserId)
+	if appErr != nil {
+		return nil, false, appErr
+	}
```

Comment:
There's a few things here that we should work on, but it will take a while to explain. Bear with me. Please don't take this as criticism of your ability at all, it's just things we all have to learn -- imagine everything is bracketed with "this is something you learn after working with Go and how we write plugins". :)

1. If you're creating a new item,  the `command.go` doesn't need to know about everything after that. Its job is to call the APIs and get things moving, but it doesn't need to know what each user has in their list, and it shouldn't be inserting todos into lists itself.
1. The List manager should probably be an API that the command calls, and the API will return an error if it failed, or nothing otherwise. something like `listManager.Add(toUserId, fromUserId, message)`.
1. The List manager will do all the "business logic" of taking a laded list (from the store), inserting todos into lists, and then telling the store to save the list. It should call the store. 
1. The store know how to load and save lists, it shouldn't know anything about what goes in the list, or how to manipulate the lists. It should have implement the `ListStore` interface, which is defined in the List Manager's file (because the List Manager is the struct that will use the ListStore to load/save lists).
1. The plugin is in charge of wiring these together. It will call NewListStore (passing in the plugin.API and maybe the plugin.Helpers) to get an instance of the store. It will pass this store into the List Manager, along with anything else the List Manager needs.
1. All of the queueing/dequeueing etc. should happen in the List Manager -- the plugin doesn't need to know about that.

I realize a lot of this design was done before you touched it, but it's good to improve the design as we go. :) 

I'm seeing if I can give you access to one of our repos that gives a good example of this.

Reply:
You are completely right. I was struggling to understand how to do the new architecture (no states, but inferred by their list) and adding the lists to simplify a part of the code that I got into really bad code. Now that I have everything more settled in my head, I think I can do better.

I will try my best to do all the refactoring using your comments. Thanks!